tests/test_network.py

```diff
@@ -124,6 +134,159 @@ def test_add_objective(factory, client):
     assert dataset.objective_key == objective.key
 
 
+@pytest.mark.parametrize('asset_name', [
+    'objective',
+    'algo',
+    'aggregate_algo',
+    'composite_algo',
+])
+def test_metadata(factory, client, asset_name):
+    method_create = getattr(factory, f"create_{asset_name}")
+    method_add = getattr(client, f"add_{asset_name}")
+
+    # add an asset with metadata
+    spec = method_create(metadata={"foo": "bar"})
+    asset = method_add(spec)
+
+    assert asset.metadata == {"foo": "bar"}
+
+    # add asset with metadata set to None
+    spec = method_create(metadata=None)
+    asset = method_add(spec)
+
+    assert asset.metadata == {}
+
+    # add asset with metadata set to an empty dict
+    spec = method_create(metadata={})
+    asset = method_add(spec)
+
+    assert asset.metadata == {}
+
+    spec = method_create(metadata='foo')
+    asset = method_add(spec)
+
+    assert asset.metadata == {}
+
+    spec = method_create(metadata={'foo': {"bar": "foo"}})
+    with pytest.raises(ValueError):
+        method_add(spec)
+
+    too_long = 'foo' * 40
+    # raise an error if the key is too long (more than 50 chars)
+    spec = method_create(metadata={too_long: "bar"})
+    with pytest.raises(substra.exceptions.InvalidRequest):
+        method_add(spec)
+
+    # raise an error if the value is too long (more than 100 chars)
+    spec = method_create(metadata={"foo": too_long})
+    with pytest.raises(substra.exceptions.InvalidRequest):
+        method_add(spec)
+
+
+@pytest.mark.parametrize('asset_name, algo_type', [
+    ('traintuple', 'algo'),
+    ('composite_traintuple', 'composite_algo'),
+])
+def test_metadata_traintuple(factory, client, asset_name, algo_type, default_dataset):
```

Comment:
This test creates a lot of tuples. These tuples should be canceled if possible and the test should not end until the tuples are either done or canceled. Otherwise the execution of tuples in other tests might be delayed by these ones and timeout.